maptivist/src/main/java/com/google/sps/servlets/MarkerServlet.java

```diff
@@ -0,0 +1,206 @@
+        String title = request.getParameter("title");
+        String longitude = request.getParameter("longitude"));
+        String latitude = request.getParameter("latitude"));
+        String description = request.getParameter("description");
+        Boolean votes = Boolean.parseBoolean(request.getParameter("upvotes"));
+
+        Set<String> linkSet = new HashSet<String>(Arrays.asList(request.getParameterValues("links")));
+        String links = createLinkString(linkSet);
+        Set<String> categorySet = new HashSet<String>(Arrays.asList(request.getParameterValues("category")));
+        String categories = createCategoriesString(categorySet);
+        ArrayList<String> flagSet = new ArrayList<>(Arrays.asList(request.getParameterValues("flags")));
+        String flags = createFlagString(flagSet);
+ 
+        if (validateInput(title, longitude, latitude, description, links)) {  
+            Entity markerEntity = new Entity("Marker");
+            markerEntity.setProperty("title", title);
+            markerEntity.setProperty("description", description);
+            markerEntity.setProperty("longitude", longitude);
+            markerEntity.setProperty("latitude", latitude);
+            markerEntity.setProperty("flags", flags);
+            markerEntity.setProperty("links", links);
+            markerEntity.setProperty("category", categories);
+            markerEntity.setProperty("votes", votes);
+
+ 
+            DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+            datastore.put(markerEntity);
+ 
+            response.sendRedirect("/maps.html");
+        } else {
+            response.sendRedirect("/maps.html");
+        }
+    }
+ 
+    private ArrayList<Marker> markers getMarkers(HttpServletRequest request){
+        ArrayList<Marker> markers = new ArrayList<>();
+ 
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        Query query = new Query("Marker");
+        PreparedQuery results = datastore.prepare(query);
+ 
+        for (Entity entity : results.asIterable()) {
+            String title = (String) entity.getProperty("title");
+            String description = (String) entity.getProperty("description");
+            Double longitude = Double.parseDouble((String) entity.getProperty("longitude"));
+            Double latitude = Double.parseDouble((String) entity.getProperty("latitude"));
+            Set<String> links = createLinkObject((String) entity.getProperty("links"));
+            ArrayList<String> flags = createFlagObject((String) entity.getProperty("flags"));
+        
+            Marker marker = new Marker(title, description, latitude, longitude, links, categories);
+            markers.add(marker);
+        }
+        return markers;
+    }
+ 
+    private Boolean validateInput(String title, String longitude, String latitude, String description, String links) {
```

Comment:
The Marker classes constructor is exactly the same as this validate input method, so I don't think this is necessary. The marker can't be made if specific components are empty.